File: java/src/main/java/org/licenselynx/LicenseLynx.java
/**
 * SPDX-FileCopyrightText: Copyright 2025 Siemens AG
 * SPDX-License-Identifier: BSD-3-Clause
 */
package org.licenselynx;

import javax.annotation.CheckForNull;

import javax.annotation.Nonnull;

import java.util.Map;


/**
 * LicenseLynx class to map license names to their corresponding data from a JSON file.
 */
public final class LicenseLynx
{

    // Private constructor to prevent instantiation
    private LicenseLynx()
    {
    }



    /**
     * Maps the given license name to its corresponding LicenseObject.
     *
     * @param pLicenseName the name of the license to map
     * @return the license data as a LicenseObject, or null if not found
     */
    @CheckForNull
    public static LicenseObject map(@Nonnull final String pLicenseName)
    {
        LicenseMapSingleton licenseMapSingleton = LicenseMapSingleton.getInstance();
        LicenseMap licenseMap = licenseMapSingleton.getLicenseMap();

        String licenseNameNormalized = QuotesHandler.normalizeQuotes(pLicenseName);
        return licenseMap.getCanonicalLicenseMap().get(licenseNameNormalized);
    }



    /**
     * Maps the given license name to its corresponding LicenseObject.
     * It also searches through the risky license mappings, when the boolean value is set.
     *
     * @param pLicenseName the name of the license to map
     * @param pRisky boolean flag to enable risky mappings
     * @return the license data as a LicenseObject, or null if not found
     */
    @CheckForNull
    public static LicenseObject map(@Nonnull final String pLicenseName, final boolean pRisky)
    {
        LicenseMapSingleton licenseMapSingleton = LicenseMapSingleton.getInstance();
        LicenseMap licenseMap = licenseMapSingleton.getLicenseMap();

        String licenseNameNormalized = QuotesHandler.normalizeQuotes(pLicenseName);
        LicenseObject licenseObject = licenseMap.getCanonicalLicenseMap().get(licenseNameNormalized);

        if (licenseObject == null && pRisky) {
            licenseObject = licenseMap.getRiskyLicenseMap().get(licenseNameNormalized);
        }

        return licenseObject;
    }



    /**
     * Maps the given license name to its corresponding LicenseObject using an extra license map.
     *
     * @param pLicenseName the name of the license to map
     * @param pExtra the extra license map identifier, or null for standard mapping
     * @return the license data as a LicenseObject, or null if not found
     */
    @CheckForNull
    public static LicenseObject map(@Nonnull final String pLicenseName, @CheckForNull final Extra pExtra)
    {
        if (pExtra == null)
        {
            return map(pLicenseName);
        }

        LicenseMapSingleton licenseMapSingleton = LicenseMapSingleton.getInstance();
        LicenseMap licenseMap = licenseMapSingleton.getLicenseMap();

        String licenseNameNormalized = QuotesHandler.normalizeQuotes(pLicenseName);
        Map<String, LicenseObject> extraMap = licenseMap.getMap(pExtra);

        if (extraMap == null)
        {
            return null;
        }

        return extraMap.get(licenseNameNormalized);
    }
}

File: python/licenselynx/licenselynx.py
#
# SPDX-FileCopyrightText: Copyright 2025 Siemens AG
# SPDX-License-Identifier: BSD-3-Clause
#
import sys
from typing import Optional
from licenselynx.license_object import LicenseObject
from licenselynx.license_map_singleton import _LicenseMapSingleton
from licenselynx.quotes_handler import _QuotesHandler
from licenselynx.extra import Extra


class LicenseLynx:

    @staticmethod
    def map(license_name: str, risky: bool = False, extra: Optional[Extra] = None) -> Optional[LicenseObject]:
        """
        Maps license name to the canonical license identifier
        :param license_name: string of a license name
        :param risky: enable risky mappings
        :param extra: enable extra mappings for a specific company/org
        :return: LicenseObject with the canonical license identifier and source, None if no license is found,
        or throws an exception if a runtime error occurs
        """
        try:
            license_name = _QuotesHandler().normalize_quotes(license_name)
            instance = _LicenseMapSingleton()

            license_object: Optional[LicenseObject] = instance.merged_data.get_canonical_map().get(license_name)

            if not license_object and risky:
                license_object = instance.merged_data.get_risky_map().get(license_name)

            if not license_object and extra:
                extra_map = instance.merged_data.get_map(extra)
                if extra_map:
                    license_object = extra_map.get(license_name)

            return license_object
        except Exception as e:
            raise e.with_traceback(sys.exc_info()[2])
